src/services/DomainModel.ts

```diff
@@ -0,0 +1,48 @@
+// Do not assume you can reach every node, sensor, or reading starting from a
+// gateway you're looking at. Be mindful that a Query is probably in effect
+// which has limited the extent to which we have populated all the objects into
+// memory. For example, if you're working with recent data from the  'Black
+// Pearl' node you might have info about the Gateway, but just the one
+// node, and a few days of readings. You'll find `Unknown`s at the points in the
+// tree that we don't care about in the context of an active Query.
+
+export interface Unknown {
+  // Unknown means you've reached an unexplored part of the tree, not a dead end.
+  explored: false;
+}
+
+export interface Query {
+  params: QueryParam[]; // TBD what a QueryParam is
+  results: Gateway[] | Node[] | Sensor[] | Reading[];
+}
+
+export interface Gateway {
+  gatewayID: string;
+  nodes: Node[] | Unknown;
+}
+
+export interface Node {
+  nodeID: string;
+  gateway: Gateway | null | Unknown;
+  sensors: Sensor[] | Unknown;
+  name: string | Unknown;
+  location: string | Unknown;
+  description: string | Unknown;
+}
+
+export interface Sensor {
```

Comment:
Surprised to see the concept of a sensor here since we didn’t have one in the DB model. Is the idea for the `Sensor` to be a logical place to declare a `schema`?

I’m cool with this, but we probably want to think through how users will initialize these objects, since we can’t derive sensors from Notehub events directly.

Reply:
Yes, it was surprising to me, but this is a more restrained and concise definition for a sensor. Later we may have an Actuator too connected to an input UI component (e.g. button, dial etc.)

Reply:
It felt like we should have something to represent a `Sensor` since that's basically the main feature of the program. Would you rather we rename `ReadingSchema` to `Sensor` or `SensorType`?

A `Sensor` is basically the cross product between Nodes and ReadingSchemas.

But we could also do this: 

```ts
export interface Node {
  sensors: Map<ReadingSchema, Set<Reading> | Unknown>> | Unknown;
  ...
}

export interface ReadingSchema {
  measure: string; // Temperature, etc.
  name: string; // Outdoor Temperature, etc.
  ...
}

export interface Reading {
  value: JSONValue;
  timestamp: Date;
  ...
}
```

I think that's what Mat and I started to do but it felt like it was getting too complicated not having a `Sensor` type.


Here's what it'd look like if we renamed `ReadingSchema` to `SensorSchema`

```ts
export interface Node {
  sensors: Map<SensorSchema, Set<Reading> | Unknown>> | Unknown;
  ...
}

export interface SensorSchema {
  measure: string; // Temperature, etc.
  name: string; // Outdoor Temperature, etc.
  ...
}

export interface Reading {
  value: JSONValue;
  timestamp: Date;
  ...
}
```

Reply:
I would say one is a class of sensor (a thermometer), and the other is an instance  of a Sensor - `Sensor` should be the latter - very specific. It is the difference between a thermometer as a concept vs the thermometer you have in your mouth taking your temperature. 

A sensor is simply something we can read which has a known constrained type of data that it reports each time. Even the word Sensor is quite limiting here compared to the generality of the concept, but it fits our domain. It's good for our needs here.

Reply:
ReadingSchema is now SensorType